protein_lm/utils/fim_collator.py

```diff
@@ -0,0 +1,88 @@
+from transformers import DataCollatorForLanguageModeling
+from protein_lm.tokenizer.tokenizer import AptTokenizer
+from typing import Any, Callable, Dict, List, NewType, Optional, Tuple, Union
+from torch.utils.data import Dataset, DataLoader
+import torch
+import random
+from collections.abc import Mapping
+from transformers import BertTokenizer
+import numpy as np
+
+class FimDataCollator(DataCollatorForLanguageModeling):
+    def __init__(self,
+        tokenizer=AptTokenizer,     
+        min_span_len: int = 1, 
+        max_span_len: int = 10,
+        fim_freq: float = 0.5, 
+        mlm_probability=0.15,
+        mlm=False,
+        **kwargs ):
+
+        super().__init__(tokenizer=tokenizer,mlm_probability=mlm_probability, **kwargs)
+        self.fim_freq = fim_freq
+        self.tokenizer=AptTokenizer()
+        self.min_span_len=  min_span_len
+        self.max_span_len= max_span_len
+        self.mlm=mlm
+
+    def fim_transform(self, inputs):
+        max_span_len = min(self.max_span_len, inputs.shape[-1])
+        if self.min_span_len == max_span_len:
+          span_len=self.min_span_len
+        else:
+          span_len=random.randint(self.min_span_len, max_span_len)
+        span_start=random.randint(0,inputs.shape[-1]-span_len)
+        span_end=span_start + span_len
+        middle_span_token = torch.tensor([self.tokenizer.middle_span_id], dtype=torch.long, device=inputs.device)
+        end_span_token = torch.tensor([self.tokenizer.end_span_id], dtype=torch.long, device=inputs.device)
+        
+        fim_transformation = [                 
+            inputs[ :span_start],             # <cls>+ Tokens before the masked span
+            middle_span_token,                # <middle_span>
+            inputs[ span_end:],               # Tokens after the masked span + <eos>
+            inputs[ span_start:span_end], #tokens in middle
+            end_span_token                      # <end_span>
+        ]
+
+        inputs=torch.cat(fim_transformation)
+        return inputs
```

Comment:
This seems right to me, can you run the tests I did for the FIM transform function I had on this too?

Reply:
I ran the tests on the FimDataCollator and all tests pass! I added them to the fim_tests.py file.

Reply:
Awesome!